pkg/cmd/cli/cmd/rollout/cancel.go

```diff
@@ -0,0 +1,202 @@
+	kubectl.AddJsonFilenameFlag(cmd, &opts.Filenames, usage)
+	kcmdutil.AddRecursiveFlag(cmd, &opts.Recursive)
+	return cmd
+}
+
+func (o *CancelOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, out io.Writer, args []string) error {
+	if len(args) == 0 && len(o.Filenames) == 0 {
+		return kcmdutil.UsageError(cmd, cmd.Use)
+	}
+
+	o.Mapper, o.Typer = f.Object(false)
+	o.Encoder = f.JSONEncoder()
+	o.Out = out
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	_, _, o.Clientset, err = f.Clients()
+	if err != nil {
+		return err
+	}
+
+	r := resource.NewBuilder(o.Mapper, o.Typer, resource.ClientMapperFunc(f.ClientForMapping), f.Decoder(true)).
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, o.Recursive, o.Filenames...).
+		ResourceTypeOrNameArgs(true, args...).
+		ContinueOnError().
+		Latest().
+		Flatten().
+		Do()
+	err = r.Err()
+	if err != nil {
+		return err
+	}
+
+	o.Infos, err = r.Infos()
+	return err
+}
+
+func (o CancelOptions) Run() error {
+	allErrs := []error{}
+	for _, info := range o.Infos {
+		config, ok := info.Object.(*deployapi.DeploymentConfig)
+		if !ok {
+			allErrs = append(allErrs, kcmdutil.AddSourceToErr("cancelling", info.Source, fmt.Errorf("expected deployment configuration, got %T", info.Object)))
+		}
+		if config.Spec.Paused {
+			allErrs = append(allErrs, kcmdutil.AddSourceToErr("cancelling", info.Source, fmt.Errorf("unable to cancel paused deployment %s/%s", config.Namespace, config.Name)))
+		}
+
+		cancelled := false
+		mapping, err := o.Mapper.RESTMapping(kapi.Kind("ReplicationController"))
+		if err != nil {
+			return err
+		}
+
+		mutateFn := func(rc *kapi.ReplicationController) {
+			if deployutil.IsDeploymentCancelled(rc) {
+				kcmdutil.PrintSuccess(o.Mapper, false, o.Out, info.Mapping.Resource, info.Name, false, "already cancelled")
+				return
+			}
+
+			patches := set.CalculatePatches([]*resource.Info{&resource.Info{Object: rc, Mapping: mapping}}, o.Encoder, func(rcInfo *resource.Info) (bool, error) {
+				deployment, ok := rcInfo.Object.(*kapi.ReplicationController)
+				if !ok {
+					return false, nil
+				}
+				deployment.Annotations[deployapi.DeploymentCancelledAnnotation] = deployapi.DeploymentCancelledAnnotationValue
+				deployment.Annotations[deployapi.DeploymentStatusReasonAnnotation] = deployapi.DeploymentCancelledByUser
+				return true, nil
+			})
+
+			if len(patches) == 0 {
+				kcmdutil.PrintSuccess(o.Mapper, false, o.Out, info.Mapping.Resource, info.Name, false, "already cancelled")
+				return
+			}
+
+			_, err := o.Clientset.ReplicationControllers(rc.Namespace).Patch(rc.Name, kapi.StrategicMergePatchType, patches[0].Patch)
+			if err != nil {
+				allErrs = append(allErrs, kcmdutil.AddSourceToErr("cancelling", info.Source, err))
+				return
+			}
+			kcmdutil.PrintSuccess(o.Mapper, false, o.Out, info.Mapping.Resource, info.Name, false, "cancelling")
+			cancelled = true
```

Comment:
This will change to true if the latest rollout for a DC is cancelled but what if you have an additional DC with the latest rollout still ongoing?